text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+set of changes that absolutely *cannot* be made without a semver bump.
+
+Along the way, it also discusses some interactions with potential language
+features that can help mitigate pain for non-breaking changes.
+
+The RFC covers only API issues; other issues related to language features,
+lints, type inference, command line arguments, Cargo, and so on are considered
+out of scope.
+
+# Motivation
+
+Both Rust and its library ecosystem have adopted [semver](http://semver.org/), a
+technique for versioning platforms/libraries partly in terms of the effect on
+the code that uses them. In a nutshell, the versioning scheme has three components::
+
+1. **Major**: must be incremented for changes that break client code.
+2. **Minor**: incremented for backwards-compatible feature additions.
+3. **Patch**: incremented for backwards-compatible bug fixes.
+
+[Rust 1.0.0](http://blog.rust-lang.org/2015/02/13/Final-1.0-timeline.html) will
+mark the beginning of our
+[commitment to stability](http://blog.rust-lang.org/2014/10/30/Stability.html),
+and from that point onward it will be important to be clear about what
+constitutes a breaking change, in order for semver to play a meaningful role. As
+we will see, this question is more subtle than one might think at first -- and
+the simplest approach would make it effectively impossible to grow the standard
+library.
+
+The goal of this RFC is to lay out a comprehensive policy for what *must* be
+considered a breaking API change from the perspective of semver, along with some
+guidance about non-semver-breaking changes.
+
+# Detailed design
+
+For clarity, in the rest of the RFC, we will use the following terms:
+
+* **Major change**: a change that requires a major semver bump.
+* **Minor change**: a change that requires only a minor semver bump.
+* **Breaking change**: a change that, *strictly speaking*, can cause downstream
+  code to fail to compile.
+
+What we will see is that in Rust today, almost any change is technically a
+breaking change. For example, given the way that globs currently work, *adding
+any public item* to a library can break its clients (more on that later). But
+not all breaking changes are equal.
+
+So, this RFC proposes that **all major changes are breaking, but not all breaking
+changes are major.**
+
+## Overview
+
+### Principles of the policy
+
+The basic design of the policy is that **minor changes should require at most a
+few local *annotations* to the code you are developing, and in principle no
+changes to your dependencies.**
+
+In more detail:
+
+* Minor changes should require at most minor amounts of work upon upgrade. For
+  example, changes that may require occasional type annotations or use of UFCS
+  to disambiguate are not automatically "major" changes. (But in such cases, one
+  must evaluate how widespread these "minor" changes are).
+
+* In principle, it should be possible to produce a version of the code for any
```

Comment:
A slightly weaker way of phrasing this would be "given knowledge of any change, it should be possible to write a version of your code that works both before and after that change". It seems to me like that phrasing is significantly more important to guarantee than the version written, since it means libraries won't be "forced" to abandon old versions of Rust to support new versions.